lib/libuzfs/libuzfs_attr.c

```diff
@@ -0,0 +1,690 @@
+	sa_handle_destroy(sa_hdl);
+	if (txg != NULL) {
+		*txg = tx->tx_txg;
+	}
+	dmu_tx_commit(tx);
+	return (0);
+}
+
+static int
+libuzfs_get_nvlist_from_handle(const sa_attr_type_t *sa_tbl,
+    nvlist_t **nvl, sa_handle_t *sa_hdl)
+{
+	int xattr_size;
+	int err = sa_size(sa_hdl, sa_tbl[UZFS_XATTR], &xattr_size);
+	if (err != 0) {
+		return (err);
+	}
+
+	char *obj = vmem_alloc(xattr_size, KM_SLEEP);
+	err = sa_lookup(sa_hdl, sa_tbl[UZFS_XATTR], obj, xattr_size);
+	if (err != 0) {
+		return (err);
+	}
+
+	err = nvlist_unpack(obj, xattr_size, nvl, KM_SLEEP);
+	vmem_free(obj, xattr_size);
+	return (err);
+}
+
+static int
+libuzfs_kvattr_set_nvlist(nvlist_t *nvl, int err_check_sa,
+    int *err_check_sa_enough, const char *name,
+    const char *value, size_t value_size, size_t *xattr_sa_size)
+{
+	int err = 0;
+	if (value == NULL && err_check_sa == 0) {
+		err = nvlist_remove(nvl, name, DATA_TYPE_BYTE_ARRAY);
+	} else if (value != NULL) {
+		err = nvlist_add_byte_array(nvl, name,
+		    (uchar_t *)value, (uint_t)value_size);
+	}
+	if (err != 0) {
+		return (err);
+	}
+
+	err = nvlist_size(nvl, xattr_sa_size, NV_ENCODE_XDR);
+	if (err != 0) {
+		return (err);
+	}
+
+	if (*xattr_sa_size > DXATTR_MAX_SA_SIZE ||
+	    *xattr_sa_size > SA_ATTR_MAX_LEN) {
+		*err_check_sa_enough = EFBIG;
+		err = nvlist_remove(nvl, name, DATA_TYPE_BYTE_ARRAY);
+		if (err != 0 && value != NULL) {
+			return (err);
+		}
+		return (nvlist_size(nvl, xattr_sa_size, NV_ENCODE_XDR));
+	}
+	return (err);
+}
+
+static void
+libuzfs_save_xattr_dir(sa_handle_t *sa_hdl, libuzfs_dataset_handle_t *dhp,
+    uint64_t xattr_obj, uint64_t value_obj, const char *name,
+    const char *value, size_t size, dmu_tx_t *tx)
+{
+	sa_attr_type_t *sa_tbl = dhp->uzfs_attr_table;
+	objset_t *os = sa_hdl->sa_os;
+	int dnodesize = dmu_objset_dnodesize(os);
+	if (xattr_obj == DMU_NEW_OBJECT) {
+		xattr_obj = zap_create_dnsize(os, DMU_OT_DIRECTORY_CONTENTS,
+		    DMU_OT_NONE, 0, dnodesize, tx);
+		VERIFY0(sa_update(sa_hdl, sa_tbl[UZFS_DXATTR],
+		    &xattr_obj, 8, tx));
+	}
+
+	if (value_obj == DMU_NEW_OBJECT) {
+		libuzfs_create_inode_with_type_impl(dhp, &value_obj,
+		    B_FALSE, INODE_FILE, tx);
+		VERIFY0(zap_update(os, xattr_obj, name,
+		    8, 1, &value_obj, tx));
+	}
+
+	dmu_write(os, value_obj, 0, size, value, tx);
+
+	sa_handle_t *value_obj_hdl = NULL;
+	VERIFY0(sa_handle_get(os, value_obj, NULL,
+	    SA_HDL_PRIVATE, &value_obj_hdl));
+	VERIFY0(sa_update(value_obj_hdl,
+	    sa_tbl[UZFS_SIZE], &size, 8, tx));
+	sa_handle_destroy(value_obj_hdl);
+}
+
+int
+libuzfs_inode_set_kvattr(libuzfs_dataset_handle_t *dhp, uint64_t ino,
+    const char *name, const char *value, uint64_t size,
+    int flags, uint64_t *txg)
+{
+	// 1. this name is in sa:
+	//    1.1 sa has enough space for new value, just put in
+	//    1.2 no enough space, delete from sa and put in dir
+	// 2. in dir
+	//    2.1 sa enough space, delete from dir and put in sa
+	//    2.2 not enough, put in dir
```

Comment:
could we just use a zap object to store "nv" if sa is not big enough, rather than use a dir?

The overhead of zap object should be less than dir.

There is one limitation for zap object is that, the max value length is 8KB, and the dir object has no such limitation. But IMO 8KB is good enough for kvattr, as the max value length of xattr for NFS is 1KB, and 8KB should be good enough for mds internal attribute.

So at least we can add a zap object between sa and dir.